### PR TITLE
update agent to create template

### LINKED ISSUE #58: HTML Template report
We should use a template engine like PUG to create human readable versions of our reports.

### PR DESCRIPTION
**Pull request by chrisperez626**
Closes #58 
The template is very boilerplate and does not show all of the data. However, all of it should be available to use in the template. We just need to edit the HTML file to show it.

### CODE REVIEW

File: src/verinfast/utils/html_template.py
Comment (aylusltd) [inline src/verinfast/utils/html_template.py:1]: can't we save this just as an hbs file?

File: src/verinfast/agent.py [hunk 1]
```diff
@@ -667,6 +678,14 @@ def main():
     agent = Agent()
     try:
         agent.scan()
+        # size_file = json.load(os.path.join(agent.config.output_dir, repo_name + ".sizes.json"))
+        with TemporaryDirectory() as tmpdir:
```
Comment (aylusltd) [inline src/verinfast/agent.py:682]: normally we do with open(...) as file: instead of having a separate close line

File: src/verinfast/agent.py [hunk 2]
```diff
@@ -7,13 +7,16 @@
 import re
 import shutil
 import subprocess
+from tempfile import TemporaryDirectory
 import traceback
 from uuid import uuid4
 
 import httpx
+from jinja2 import Environment, FileSystemLoader
```
Comment (aylusltd) [inline src/verinfast/agent.py:15]: need to add to requirements.txt

File: src/verinfast/agent.py [hunk 3]
```diff
@@ -7,13 +7,16 @@
 import re
 import shutil
 import subprocess
+from tempfile import TemporaryDirectory
```
Comment (aylusltd) [inline src/verinfast/agent.py:10]: seems unnecessary...

File: src/verinfast/agent.py [hunk 4]
```diff
@@ -667,6 +678,14 @@ def main():
     agent = Agent()
     try:
         agent.scan()
+        # size_file = json.load(os.path.join(agent.config.output_dir, repo_name + ".sizes.json"))
+        with TemporaryDirectory() as tmpdir:
+            tmpfile = open(f"{tmpdir}/results.html", "w")
```
Comment (aylusltd) [inline src/verinfast/agent.py:683]: having the string in html would just skip this whole step

**Review by aylusltd — CHANGES_REQUESTED** [review]

File: src/verinfast/agent.py [hunk 5]
```diff
@@ -364,12 +370,13 @@ def parseRepo(self, path: str, repo_name: str):
...
                     f.write(json.dumps(filelist, indent=4))
 
+                template_definintion["filelist"] = filelist
                 # Calling modernmetric with subprocess works, but we might want to call
                 # Modernmetric directly, ala lines 91-110 from modernmetric main
                 with open(stats_output_file, 'w') as f:
                     with open(stats_error_file, 'w') as e:
                         subprocess.check_call(["modernmetric", f"--file={stats_input_file}"], stdout=f, stderr=e, encoding='utf-8')
-
+                template["stats"] = json.load(stats_output_file)
```
Comment (aylusltd) [inline src/verinfast/agent.py:379]: You mean template_definition right?

File: templates/results.j2 [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        {% for result in gitfindings.results%}
+            <tr>
+                <td>{{result|escape}}</td>
+            </tr>
+        {% endfor %}
+        </table>
+        {% else %}
+        {% endif %}
+    </body>
+</html>
```
Comment (aylusltd) [inline templates/results.j2:63]: Please add new line at end of file

File: templates/results.j2 [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    <head>
+        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.3.1/dist/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
+    <head>
+    <body>
+        <table class="table table-striped">
+        <tr>
+            <th>Name</th>
+            <th>Path</th>
+        </tr>
+        {% for file in filelist%}
```
Comment (aylusltd) [inline templates/results.j2:12]: We have sizes and lines of code too

**Review by aylusltd — CHANGES_REQUESTED** [review]

File: src/verinfast/agent.py (hunk 5 above)
Comment (chrisperez626) [reply src/verinfast/agent.py]: There's still an issue here when trying to add stats to the template definition. I'm not sure why `json.load(stats_output_file)` is loading an empty dictionary after `modernmetric` is run.

File: templates/results.j2 [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        {% if stats %}
+        <table class="table table-striped">
+            <tr>
+                <th>File</th>
+                <th>Data</th>
+            </tr>
+            {% for k, v in stats.files.items()%}
+                <tr>
+                    <td>{{k|escape}}</td>
+                    <td>{{v|escape}}</td>
```
Comment (aylusltd) [inline templates/results.j2:71]: that is a lazy way to skip labeling columns, but I approve.

File: templates/results.j2 [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        </table>
+        {% else %}
+        {% endif %}
+        {% if sizes %}
+        <table class="table table-striped">
+            <tr>
+                <th>File</th>
+                <th>Size</th>
+            </tr>
+            {% for k,v in sizes.files.items()%}
```
Comment (aylusltd) [inline templates/results.j2:83]: couldn't you have merged these two together at least?

File: templates/results.j2 [hunk 10]
```diff
@@ -0,0 +1,93 @@
+<!DOCTYPE html>
+    <html lang="en">
+    <head>
+        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.3.1/dist/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
+    <head>
+    <body>
+        <table class="table table-striped">
+            <tr>
+                <th>Name</th>
+                <th>Path</th>
```
Comment (aylusltd) [inline templates/results.j2:10]: should merge this with the other two, each file just needs one row.

**Review by aylusltd — CHANGES_REQUESTED** [review]
Make one table for all files, and put it at the top, followed by semgrep findings. github commits should also be limited to most recent hundred or page will crash.

File: templates/results.j2 [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            </tr>
+            <tr>
+                <td>Current directory data</td>
+                <td>{{sizes.files.pop(".")|escape}}</td>
+            </tr>
+            {% for (file, file_size_data) in zip(filelist, sizes.files.values())%}
+                <tr>
+                    <td>{{file.name|escape}}</td>
+                    <td>{{file_size_data|escape}}</td>
+                    {#<td>{{file_stat_data|escape}}</td>#}
```
Comment (aylusltd) [inline templates/results.j2:20]: should uncomment this and add a description above

File: templates/results.j2 [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            </tr>
+            {% for result in gitfindings.results%}
+                <tr>
+                    <td>{{result|escape}}</td>
+                </tr>
+            {% endfor %}
+        </table>
+        {% else %}
+        {% endif %}
+    </body>
```
Comment (aylusltd) [inline templates/results.j2:64]: add dependency table too

File: src/verinfast/agent.py [hunk 13]
```diff
@@ -667,6 +676,11 @@ def main():
     agent = Agent()
     try:
         agent.scan()
+        with open(f"{agent.config.output_dir}/results.html", "w") as f:
+            jinja_env = Environment(loader=FileSystemLoader("templates/"))
+            jinja_env.globals.update(zip=zip)
+            output = jinja_env.get_template("results.j2").render(template_definition)
+            f.write(output)
```
Comment (SeanTConrad) [inline src/verinfast/agent.py:683]: Getting this error when I run this:
`Traceback (most recent call last):
  File "/home/stconrad/.local/bin/verinfast", line 8, in <module>
    sys.exit(main())
  File "/home/stconrad/chris/verinfast/src/verinfast/agent.py", line 688, in main
    raise e
  File "/home/stconrad/chris/verinfast/src/verinfast/agent.py", line 682, in main
    output = jinja_env.get_template("results.j2").render(template_definition)
  File "/home/stconrad/.local/lib/python3.10/site-packages/jinja2/environment.py", line 1010, in get_template
    return self._load_template(name, globals)
  File "/home/stconrad/.local/lib/python3.10/site-packages/jinja2/environment.py", line 969, in _load_template
    template = self.loader.load(self, name, self.make_globals(globals))
  File "/home/stconrad/.local/lib/python3.10/site-packages/jinja2/loaders.py", line 126, in load
    source, filename, uptodate = self.get_source(environment, name)
  File "/home/stconrad/.local/lib/python3.10/site-packages/jinja2/loaders.py", line 218, in get_source
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: results.j2`

Comment (chrisperez626) [reply src/verinfast/agent.py:683]: What is your output for `pwd`?

File: tests/test_template.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    config.output_dir = results_dir
+    print(agent.config.output_dir)
+    agent.config = config
+    agent.config.shouldUpload = True
+    agent.debug = DebugLog(path=agent.config.output_dir, debug=False)
+    agent.log = agent.debug.log
+    print(agent.debug.logFile)
+    agent.scan()
+    assert Path(results_dir).exists()
+    assert results_dir.joinpath("results.html").exists()
```
Comment (aylusltd) [inline tests/test_template.py:35]: please add a blank line at the end of every file and make sure your editorconfig has that.

File: src/verinfast/agent.py (hunk 13 above)
Comment (aylusltd) [reply src/verinfast/agent.py:711]: change to abspath resolves this and covered by new unit test.

**Pull request merged**